### PR TITLE
Add PhyWhisperer backend.

### PR DESCRIPTION
**Pull request by jpcrypt**

### CODE REVIEW

File: viewsb/backends/phywhisperer.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        if len(args.pattern) > cls.MAX_PATTERN_SIZE:
+            sys.stderr.write("pattern cannot have more than %s elements\n" % cls.MAX_PATTERN_SIZE)
+            sys.exit(errno.EINVAL)
+
+        #  Return the class and leftover arguments.
+        return (args.size, args.burst, args.pattern, args.mask, args.addpattern), leftover_args
+
+
+    def __init__(self, size, burst, pattern, mask, addpattern, suppress_packet_callback=None):
+        """ Creates a new PhyWhisperer capture backend.
```
Comment (ktemkin) [inline viewsb/backends/phywhisperer.py:114]: `pw` isn't imported in this scope -- it looks like you're missing something like:

```
from phywhisperer import usb as pw
```

at the top of this method.

File: viewsb/backends/phywhisperer.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        parser.add_argument('--size', type=int, default=cls.MAX_CAPTURE_SIZE, 
+                help="capture size (0 = unlimited")
+        parser.add_argument('--pattern', type=int, nargs='+', default=[0], choices=range(0,256),
+                help="capture pattern (list of ints)")
+        parser.add_argument('--mask', type=int, nargs='+', default=[0], choices=range(0,256),
+                help="mask pattern (list of ints)")
+        parser.add_argument('--addpattern', action='store_true',
+                help="add pattern to captured data")
+        parser.add_argument('--burst', action='store_true',
+                help="add pattern to captured data")
```
Comment (ktemkin) [inline viewsb/backends/phywhisperer.py]: Looks like this argument's documentation didn't make it through. :)

File: viewsb/backends/phywhisperer.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    class ViewSBEventSink(USBEventSink):
+        """ PhyWhisperer USB event sink that submits packets for decoding. """
+
+
+        def __init__(self, backend, suppress_packet_callback=None):
+
+            # Store a reference to our parent backend, so we can submit USB data via it.
+            self.backend = backend
+
+            # Mark ourselves as having no packet pending.
```
Comment (ktemkin) [inline viewsb/backends/phywhisperer.py]: `pending_packet` doesn't seem to be used, here. (I think the same dead code exists in the OpenVizsla backend; which in turn was left over from the USBProxy backend.)

File: viewsb/backends/phywhisperer.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            # Store the callback we use to determine if we should suppress packets.
+            self.suppress_packet_callback = suppress_packet_callback
+
+
+        def _emit_packet(self, packet):
+            """ Emits a packet to the main decoder thread for analysis. """
+            self.backend.emit_packet(packet)
+
+
+        def _should_be_suppressed(self, packet):
```
Comment (ktemkin) [inline viewsb/backends/phywhisperer.py]: This is built-in to the main analyzer stack, now. Backends don't explicitly need to handle SOF filtering; so the whole "should be suppressed" mechanism doesn't need to be around. :)

**Review by ktemkin — CHANGES_REQUESTED** [review]
This code seems good (nice and minimal, too). In testing, I was able to get short captures working, but I haven't had luck with `--size 0` / continuous captures. I suspect the relevant changes to get that working well will need to be in `phywhisperer`, though, and not here.

Either way, there's at least one blocking thing that needs to be fixed (the import issue mentioned in comments) before this can be merged; and few other comments that you might want to address. Should be a quick merge after that. :)

**Comment by jpcrypt** [comment]
Thanks for the review! Yeah size=0 is work-in-progress on the PhyWhisperer side, it's intended for stream capture which isn't working yet.

File: viewsb/backends/phywhisperer.py (hunk 1 above)
Comment (jpcrypt) [reply viewsb/backends/phywhisperer.py:114]: fixed

File: viewsb/backends/phywhisperer.py (hunk 2 above)
Comment (jpcrypt) [reply viewsb/backends/phywhisperer.py]: fixed

File: viewsb/backends/phywhisperer.py (hunk 3 above)
Comment (jpcrypt) [reply viewsb/backends/phywhisperer.py]: removed

File: viewsb/backends/phywhisperer.py (hunk 4 above)
Comment (jpcrypt) [reply viewsb/backends/phywhisperer.py]: removed

**Comment by ktemkin** [comment]
Finally getting around to testing the updated changes; sorry for the delay. :)

**Comment by ktemkin** [comment]
This seems to work; so merging it as a baseline. :)

Thanks!

**Pull request merged**